Move-only: Pull wallet code out of libbitcoin_server

ryanofsky opened the pull request:
This is a move-only commit. No code is changing and the moves can be easily verified with:

```sh
git log -p -n1 --color-moved=dimmed_zebra
```

This commit moves functions and variables that wallet code depends on out of libbitcoin_server.a, so the bitcoin-wallet tool can be built without libbitcoin_server.a in #15639, and attempting to access server state from wallet code will result in link errors instead of silently broken code.

List of moves:

- `CheckTransaction` moves from `consensus/tx_verify.cpp` to `consensus/tx_check.cpp`
- `urlDecode` moves from `httpserver.cpp` to `util/url.cpp`
- `TransactionErrorString` moves from `node/transaction.cpp` to `util/error.cpp`
- `StringForFeeReason` and `FeeModeFromString` move from `policy/fees.cpp` to `util/fees.cpp`
- `incrementalRelayFee` `dustRelayFee` and `nBytesPerSigOp` move from `policy/policy.cpp` to `policy/settings.cpp`
- `SignalsOptInRBF` moves from `policy/rbf.cpp` to `util/rbf.cpp`
- `fIsBareMultisigStd` moves from `validation.cpp` to `policy/settings.cpp`
- `ConstructTransaction` `TxInErrorToJSON` and `SignTransaction` move from `rpc/rawtransaction.cpp` to `rpc/rawtransaction_util.cpp`
- `RPCTypeCheck` `RPCTypeCheckArgument` `RPCTypeCheckObj` `AmountFromValue` `ParseHashV``ParseHashO` `ParseHexV` `ParseHexO` `HelpExampleCli` and `HelpExampleRpc` move from `rpc/server.cpp` to `rpc/util.cpp`
- `AmountHighWarn` and `AmountErrMsg` move from `ui_interface.cpp` to `util/error.cpp`
- `FormatStateMessage` and `strMessageMagic` move from `validation.cpp` to `util/validation.cpp`
- `VerifyWallets` `LoadWallets` `StartWallets` `FlushWallets` `StopWallets` and `UnloadWallets` move from `wallet/init.cpp` to `wallet/node.cpp`

Empact commented:
I split this out for review here: https://github.com/Empact/bitcoin/commits/pr/link-split and came away with 2 proposed changes:

https://github.com/Empact/bitcoin/commit/3d5f9e943ae5e0da0edc55a7ba6e7d32689563f8 - `TransactionErrorString` is not used outside `rpc/util.cpp`, which includes `node/transaction.h` via `rpc/util.h`. Seems appropriate to leave it there, contra https://github.com/Empact/bitcoin/commit/7a6b74faee3b5e595ceb62d1df4c6c4cc87ac99a

https://github.com/Empact/bitcoin/commit/173d3c499f5639455293ea53b09677805f6128fe - this moves `StringFeeForReason` back as it's only used in `wallet/wallet.cpp` which includes `policy/fees.h` directly. It also makes `util/fees` independent of `policy/fees`, which seems appropriate re. layering.

ryanofsky commented:
@Empact, I can use your branch with split up commits if other people think that is easier to review. But I'm not sure that they will. Personally I would rather scan a single commit with `--color-moved=dimmed_zebra`, verify that it is move-only, and be able to see the whole picture than to have to look at fragmentary commits.

The two changes you propose are not good, though. They have some problems conceptually, and I think will break the followup PR #15639, although I didn't test.

_The point of this PR is to designate a directory ([`src/node/`](https://github.com/ryanofsky/bitcoin/tree/pr/link.5/src/node/)) and a library ([`libbitcoin_server.a`](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/Makefile.am#L247-L250)) which contain node-specific code, which wallet code cannot call, and qt code cannot call._ This is enforced by the linker after #15639, and should be clearly explained in the [readme](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/node/README.md) and [makefile](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/Makefile.am#L247-L250).

 Both wallet code and qt code should be able to translate error codes into strings. So [`TransactionErrorString`](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/util/error.h#L22) belongs in a place like [`util/error.h`](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/util/error.h#L22) and not underneath [`src/node/`](https://github.com/ryanofsky/bitcoin/tree/pr/link.5/src/node/). (Beyond #15639, I would like [`util/error.h`](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/util/error.h) to be a common place for defining simple error types and messages that don't pull in outside dependencies, so error values only need to be defined once, and we don't need to write unnecessary code to catch exceptions or translate errors to different types when they are passed between wallet/node/rpc/gui code boundaries.)

Similarly, both wallet code and qt code should be able to call [`StringForFeeReason`](https://github.com/ryanofsky/bitcoin/blob/pr/link.5/src/util/fees.cpp#L24-L42), which means it shouldn't be defined in [`src/policy/fees.cpp`](https://github.com/bitcoin/bitcoin/blob/abd914ed34f12a5e7773e87e0457b04bcb7d58c4/src/policy/fees.cpp#L30-L48), because that file uses mempool types and would require wallet and qt code to be linked with mempool code, which is what this PR is trying to avoid and would be an error after #15639.

jnewbery commented:
Very strong concept ACK to this and #15639. This is the natural follow-up to #10973. That PR defined a node<->wallet interface and made it so wallet code _shouldn't_ directly access node functions. This PR with #15639 ensures that wallet code _can't_ directly access node functions.

My only question is when these should be merged. This conflicts with a lot of PRs (including high-priority #15141) so perhaps it should wait. On the other hand, rebases will be trivial since this is just a code move.

251Labs commented on test/lint/lint-circular-dependencies.sh:
```diff
@@ -26,6 +27,7 @@ EXPECTED_CIRCULAR_DEPENDENCIES=(
     "qt/transactiontablemodel -> qt/walletmodel -> qt/transactiontablemodel"
     "qt/walletmodel -> qt/walletmodeltransaction -> qt/walletmodel"
     "txmempool -> validation -> txmempool"
+    "util/validation -> validation -> util/validation"
```
Is there a strong argument that justifies moving helper function `FormatStateMessage(const CValidationState &state)` to its own translation unit, whilst it depends on the translation unit it was moved from? 

I would prefer not to introduce new circular dependencies if they can be prevented.

251Labs commented on test/lint/lint-circular-dependencies.sh:
```diff
@@ -13,6 +13,7 @@ EXPECTED_CIRCULAR_DEPENDENCIES=(
     "checkpoints -> validation -> checkpoints"
     "index/txindex -> validation -> index/txindex"
     "policy/fees -> txmempool -> policy/fees"
+    "policy/policy -> policy/settings -> policy/policy"
```
Let's try not to introduce new circular dependencies. I had a quick look at #15639 and it seems that we can keep all the moved code in the `policy` translation unit and prevent the circular dependency?

jnewbery commented:
>  if other people think that is easier to review

I think it would be slightly easier to review if this was split into multiple commits, each moving a function with with justification for the move.

Sjors reviewed (commented):
tACK b35ba2c161b739c0caf9734a5e16123ed7688450

The dimmed zebra helps with checking the moves. I'm fine with either splitting this into smaller commits that explain which things are are moved and why, or just putting a list in the PR description.

I'm able to compile on macOS 10.14.3 with and without wallet.

For some reason  the x64 build in Visual Studio 2017 on Windows breaks:
<img width="1625" alt="win" src="https://user-images.githubusercontent.com/10217/54997955-4077e500-4fcd-11e9-8d62-4457b91759da.png">

AppVeyor seems happy though. I did a clean before building. Building the parent commit also breaks, although with different errors. Maybe we recently introduced a problem? cc @ken2812221 & @sipsorcery

sipsorcery commented:
@Sjors this PR builds correctly for me with VS 2017 x64 Debug and Release. Did you re-run the `msvc-autogen.py` script after pulling the changes? There have been modifications to the makefiles so the VS project files need to be re-generated.

Sjors commented:
@sipsorcery I did not. Can't test it at the moment, but that seems like a plausible explanation.

ryanofsky reviewed (commented):
Updated b35ba2c161b739c0caf9734a5e16123ed7688450 -> 9742ff8d79402607159488eed362b7eb08580114 ([pr/link.5](https://github.com/ryanofsky/bitcoin/commits/pr/link.5) -> [pr/link.6](https://github.com/ryanofsky/bitcoin/commits/pr/link.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/link.5..pr/link.6)) adding suggested comments, removing circular lint warnings.

re: https://github.com/bitcoin/bitcoin/pull/15638#pullrequestreview-218844829

> The dimmed zebra helps with checking the moves. I'm fine with either splitting this into smaller commits that explain which things are are moved and why, or just putting a list in the PR description

I added a list of moves to the PR description. A few people have requested splitting this PR up, and if someone still thinks this is a good idea, I'd urge them to at least first spend a few minutes looking over the commit with `--color-moved=dimmed_zebra`.

I don't want to break up this PR into a large number of commits, because we have lots of circular dependencies and lots of different binaries (tests, benchmarks, tools, frontends) that pull object files in with different link orders. We also build on many different platforms, so right now any commit that moves a symbol from one object file to another is basically playing russian roullete with travis and appveyor. I spent hours after I first got this PR working locally to eventually get it working on the various travis platforms and appveyor. I don't want to have to repeat this work again with a bunch of individual commits, or to blindly push a bunch of broken commits. I think splitting the commit up at best would only make it slightly easier to review with `--color-moved=dimmed_zebra`, and in actuality would just make this more intimidating and more awkward to review and maintain. I think a bitcoin contributor who literally spent 10 minutes looking at the current commit with `--color-moved=dimmed_zebra`, could fully review the PR, verifying that it is move-only and reading all the new comments.

ryanofsky commented:
John was asking questions about why different functions were moving to various libraries. I'll describe how I was thinking about this below, but I don't consider the question to be extremely important, because the libraries are already pretty scrambled up, and I'm not trying to unscramble them here. All I am trying to do is split up .cpp files that have functions called by the wallet, so wallet code that links these .cpp files doesn't pull in a bunch of node dependencies.

But just to provide reassurance that I am not messing up the libraries in some way here, here is how I am thinking about the organization:

- `libbitcoin_server.a`, `libbitcoin_wallet.a`, and `libbitcoinqt.a` should all be terminal dependencies. They should be able to depend on other symbols in other libraries, but no other libraries should depend on symbols in them (and they shouldn't depend on each other).

- `libbitcoin_consensus.a` should be a standalone library that doesn't depend on symbols in other libraries mentioned here

- `libbitcoin_common.a` and `libbitcoin_util.a` seem very interchangeable right now and mutually depend on each other. I think we should either merge them into one library, or create a new top-level `src/common/` directory complementing `src/util/`, and start to distinguish general purpose utility code (like argument parsing) from bitcoin-specific utility code (like formatting bip32 paths and using ChainParams). Both these libraries can be depended on by `libbitcoin_server.a`, `libbitcoin_wallet.a`, and `libbitcoinqt.a`, and they can depend on `libbitcoin_consensus.a`. If we want to split util and common up, as opposed to merging them together, then util shouldn't depend on libconsensus, but common should.

Over time, I think it'd be nice if source code organization reflected library organization . I think it'd be nice if all `libbitcoin_util` source files lived in `src/util`, all `libbitcoin_consensus.a` source files lived in `src/consensus`, and all `libbitcoin_server.a` code lived in `src/node` (and maybe the library was called `libbitcoin_node.a`).

jnewbery committed:
[build] Move CheckTransaction from lib_server to lib_consensus

CheckTransaction is a context-free function that does not require access
to the blockchain or mempool. Move it from src/consensus/tx_verify in
lib_server to a new unit src/consensus/tx_check in lib_consensus so that
it can be called by non-server libraries.

jnewbery committed:
[build] Move policy settings to new src/policy/settings unit

This moves the following policy settings functions and globals to a new
src/policy/settings unit in lib_server:

- `incrementalRelayFee`
- `dustRelayFee`
- `nBytesPerSigOp`
- `fIsBareMultisigStd`

These settings are only required by the node and should not be accessed
by other libraries.

jnewbery committed:
[build] Move rpc utility methods to rpc/util

Moves the following utility methods to rpc/util and moves that unit to
libbitcoin_common so they can be accessed by all libraries.

- `RPCTypeCheck`
- `RPCTypeCheckArgument`
- `RPCTypeCheckObj`
- `AmountFromValue`
- `ParseHashV``ParseHashO`
- `ParseHexV`
- `ParseHexO`
- `HelpExampleCli`
- `HelpExampleRpc`

jnewbery committed:
[build] Move rpc rawtransaction util functions to rpc/rawtransaction_util.cpp

rpc/rawtransaction.cpp moves to libbitcoin_server since it should not be
accessed by non-node libraries. The utility following utility methods
move to their own unit rpc/rawtransaction_util since they need to be
accessed by non-node libraries:

- `ConstructTransaction`
- `TxInErrorToJSON`
- `SignTransaction`

jnewbery committed:
[build] Move several units into common libraries

Moves the following units into libbitcoin_util or libbitcoin_common
since they are required by multiple libraries:

- bloom
- interfaces/handler
- merkleblock
- outputtype

jnewbery committed:
[build] Add several util units

Adds the following util units and adds them to libbitcoin_util:

- `util/url.cpp` takes `urlDecode` from `httpserver.cpp`
- `util/error.cpp` takes `TransactionErrorString` from
  `node/transaction.cpp` and `AmountHighWarn` and `AmountErrMsg` from
  `ui_interface.cpp`
- `util/fees.cpp` takes `StringForFeeReason` and `FeeModeFromString` from `policy/fees.cpp`
- `util/rbf.cpp` takes `SignalsOptInRBF` from `policy/rbf.cpp`
- 'util/validation.cpp` takes `FormatStateMessage` and `strMessageMagic` from 'validation.cpp`

jnewbery committed:
[build] Move wallet load functions to wallet/load unit

Moves the following wallet load functions to a new wallet/load unit in
the libbitcoin_wallet library. All other functions in wallet/init remain
in libbitcoin_server:

- `VerifyWallets`
- `LoadWallets`
- `StartWallets`
- `FlushWallets`
- `StopWallets`
- `UnloadWallets`

jnewbery committed:
[docs] Document src subdirectories and different libraries

jnewbery commented:
I split this PR into smaller commits to help my review: https://github.com/jnewbery/bitcoin/tree/pr15638_8 . I pushed each commit as a separate branch on my fork to be built by travis: https://travis-ci.org/jnewbery/bitcoin/builds.

I personally find this easier to review as each move is justified by its commit log.

> I'll describe how I was thinking about this below [...] here is how I am thinking about the organization

This is really useful. Would you mind opening an issue with basically the description you've given above. I think it'd be useful to have as a reference for how we think the code organization should be moving, and as a place to solicit feedback from other contributors.

jnewbery commented:
@laanwj has also expressed a desire to reorganize the source code (https://github.com/bitcoin/bitcoin/issues/15465#issuecomment-475850089):

> What kind of frustrates me, personally, is that we didn't manage to clearly separate out the consensus parts. Not to a library (which would be useful for rust-bitcoin et al), but also not even to a separate directory.

Empact commented:
Here are 2 revised proposals, related to the earlier work:

https://github.com/Empact/bitcoin/commit/c72764eccd5d769f2574bcdc39a992cab11d4048
https://github.com/Empact/bitcoin/commit/90293e6c25f691bdf2218b1c37e98f14394f48df

Based on current `pr/link`

practicalswift commented:
Concept ACK

ryanofsky commented:
Updated 9742ff8d79402607159488eed362b7eb08580114 -> fee746f72ba53642bfd1000b6cd215048ad06920 ([pr/link.6](https://github.com/ryanofsky/bitcoin/commits/pr/link.6) -> [pr/link.7](https://github.com/ryanofsky/bitcoin/commits/pr/link.7)) to John's branch [pr15638_8](https://github.com/jnewbery/bitcoin/commits/pr15638_8) plus a compile fix for the fourth commit (f12781db443d391e1f27e09bb4bc5f0148935830)

ryanofsky committed:
[build] Move AnalyzePSBT from psbt.cpp to node/psbt.cpp

psbt.cpp definitions except for AnalyzePSBT are used by the wallet and need to
be linked into the wallet binary. AnalyzePSBT is an exception in that it is not
used by the wallet, and depends on node classes like CCoinsViewCache, and on
node global variables like nBytesPerSigOp.

So AnalyzePSBT is more at home in libbitcoin_server than libbitcoin_common, and
in any case needs to be defined in a separate object file than other PSBT
utilities, to avoid dragging link dependencies on node functions and global
variables into the wallet.

ryanofsky commented:
Rebased fee746f72ba53642bfd1000b6cd215048ad06920 -> b0ba5d943100857daeb0904a01f096aad0cda697 ([pr/link.7](https://github.com/ryanofsky/bitcoin/commits/pr/link.7) -> [pr/link.8](https://github.com/ryanofsky/bitcoin/commits/pr/link.8)) due to conflict with #15508. Text merge conflict was limited to `src/Makefile.am`, but #15508 also required adding a new commit: "[build] Move AnalyzePSBT from psbt.cpp to node/psbt.cpp" (b0ba5d943100857daeb0904a01f096aad0cda697)

jnewbery commented:
tested ACK b0ba5d943100857daeb0904a01f096aad0cda697

laanwj commented:
Concept ACK

ryanofsky commented:
Rebased b0ba5d943100857daeb0904a01f096aad0cda697 -> 4d074e84a2cf419510e2920417799f62747f4b07 ([pr/link.8](https://github.com/ryanofsky/bitcoin/commits/pr/link.8) -> [pr/link.9](https://github.com/ryanofsky/bitcoin/commits/pr/link.9)) due to conflict with #15632

laanwj commented:
> libbitcoin_common.a and libbitcoin_util.a seem very interchangeable right now and mutually depend on each other

interesting—`util` is linked by:
```
bitcoind
bitcoin-cli
bitcoin-tx
bitcoin-wallet
bench/bench_bitcoin
qt/bitcoin-qt
qt/test/test_bitcoin-qt
test/test_bitcoin
```
whereas common is linked by:
```
bitcoind
bitcoin-tx
bitcoin-wallet
bench/bench_bitcoin
qt/bitcoin-qt
qt/test/test_bitcoin-qt
test/test_bitcoin
```
The only difference being `bitcoin-cli`. Yes, with this I'd say it makes sense to merge them, *unless* the current split avoids pulling in some dependencies for `bitcoin-cli` (@luke-jr might be able to comment w.r.t. gentoo).

jnewbery commented:
utACK 4d074e84a2cf419510e2920417799f62747f4b07 (checked by doing the rebase myself and verifying no difference between my branch and 4d074e84a2cf419510e2920417799f62747f4b07)

jnewbery commented:
@laanwj - do you mind moving discussion of source code organization to #15732?

laanwj commented:
utACK 4d074e84a2cf419510e2920417799f62747f4b07
